### PR TITLE
Add arabic translation

### PR DESCRIPTION
**Pull request by ahmed-0011**

### CODE REVIEW

**Comment by binarynonsense** [comment]
Thanks a lot for starting the Arabic localization file!! Seems like it wasn't based on the latest English version so I've merged it and added the missing entries with a [NEEDS TRANSLATION] tag. If you have the time, could you translate those? (there are only a few new ones).

Also, could you create a translated README.ar.md file in the doc folder?
You can use the English one as a base: https://github.com/binarynonsense/comic-book-reader/blob/master/docs/README.en.md (I understand if you don't have the time, but it would be helpful to keep the Arabic localization on par with the others)

**Comment by binarynonsense** [comment]
Here are some screenshots with the Arabic localization enabled so you can start to check if things look ok. I have no experience with Arabic or mixing right to left and left to right languages so I'll need help making sure it works as it should and how to fix any issue (I've noticed one thing, in the last screenshot, where the word version and parts of the version number, the -beta1, get shuffled but have no idea why)

image

image

image

image

**Comment by binarynonsense** [comment]
@ahmed-0011 I've tested setting the dir attribute of the whole thing to "rtl" this is what I get. Also, here are some questions and thoughts:

- Is this better in general? Things aligned to the right, including the menu bar (File, View, Tools),  AND with the window buttons (close, maximize and minimize) aligned to the left (the title seems to be aligned to the right, I think it should also be aligned to the left, next to the window buttons). From what I quickly gathered this is more how software and webs are expected to look in Arabic/RTL languages, but I may be wrong, as I said I don't have experience with this topic.
- As you can see some things don't work correctly (like the menu bar, which when an entry is selected it gets cut out to the right) and the design in general... I don't know if I'd be able to fix this (for example the menu bar is based on a package made by others, so I'd need to see if I can change it, and redesigning the UIs for rtl would also be time-consuming) or if and when I'd have the time to do such a big overhaul so:
  - would it be worth the effort? is this (but working correctly) ideally the way it should look for Arabic? Any detail you notice that is incorrect / would need changing? (for example, would it be better if top X and the bottom cancel/ok... buttons in the modals, like the ACBR info one in one screenshot, were aligned to the left and the text to the right).
  - would leaving it like the rest of the languages, left to right, but just using the Arabic translated text (like in the previous images I showed you yesterday) be a good start for know, hoping I'll manage to do it correctly some day (but no promises) or it's unusable/incorrect like that and I should wait to implement this correctly before adding Arabic to a release (it may never happen)?

image

image

image

image

**Comment by ahmed-0011** [comment]
Hi, @binarynonsense

I made a new pull request with ar.json updates. -> #112
Regarding translating the README.md file, I will try to do it as soon as I can. Probably within the next 2 weeks.

**Comment by ahmed-0011** [comment]
> @ahmed-0011 Thanks a lot for starting the Arabic localization file!! Seems like it wasn't based on the latest English version so I've merged it and added the missing entries with a [NEEDS TRANSLATION] tag. If you have the time, could you translate those? (there are only a few new ones).
> 
> Also, could you create a translated README.ar.md file in the docs folder? You can use the English one as a base: https://github.com/binarynonsense/comic-book-reader/blob/master/docs/README.en.md (I understand if you don't have the time, but it would be helpful to keep the Arabic localization on par with the others)

Hi, @binarynonsense

I made a new pull request with ar.json updates. -> #112
Regarding translating the README.md file, I will try to do it as soon as I can. Probably within the next 2 weeks.

**Comment by binarynonsense** [comment]
- Thanks a lot for updating the Arabic localization @ahmed-0011 !
- No problem, take all the time you need. When you do, add this line, translated, to the contributions section: "Arabic localization by [ahmed-0011](https://github.com/ahmed-0011)".
- Also, if you have the time, please take a look at the screenshots I posted and answer my questions about UI design and RTL languages. Hopefully you'll be able to help me solve some of my doubts and give me some hints. If you can check my previous two posts above to see if there are other things but my main worries/doubts are probably:
  - Is just applying the Arabic localization file, without any extra support for RTL languages, be at least usable/good enough for an Arabic-speaking user? (see my first set of screenshots to see how it looks and if you spot any issues/errors). If it is, I could enable Arabic in the languages list in the next stable release, if not I'd have to wait until I manage to add full support for RTL languages (which I may or may not be able to, for some of the reasons I mentioned above, and also don't know when I'll be able to dedicate the time to try doing).
  - Would full RTL support look something like the second set of screenshots I posted? (with the cropping menus, background color and icon misalignment errors fixed), or there's something missing/wrong? (there are some extra questions about that/doubts in my post above). From what I've been reading about the topic that seems to be the case, mostly mirroring most things, but as I said I have no experience so I'd need confirmation/some guidance :)

**Comment by ahmed-0011** [comment]
> (for now, I've already noticed one thing: in the last screenshot, where the word version and parts of the version number, the -beta1, get shuffled but looking at the code I have no idea why)

I think that the reason it got messed up is simply because of the mixing of two languages with different directions of writing.
In that case, it seems that numbers follow the direction of the preceding word. which in that case is Arabic since "الإصدار" is the first word.

As you can see here:

image

**Comment by ahmed-0011** [comment]
> * Is this better in general? Things aligned to the right and flowing right to left, including the menu bar (File, View, Tools) with File starting the first at the right,  AND with the window buttons (close, maximize and minimize) aligned to the left (the title "Comic Book Reader - ACBR" seems to be aligned to the right, I think it should also be aligned to the left, next to the window buttons). From what I quickly gathered this is more how software and webs are expected to look in Arabic/RTL languages, but I may be wrong, as I said I don't have experience with this topic.

Yes, this is correct behavior. menu on the right, and close/maximize/minimize on the left. For the title, I think it should be either on the left side near the buttons or in the center.

> * As you can see some things don't work correctly (like the menu bar, which when an entry is selected it gets cut out to the right) and the UI design in general breaks in some ways (like the background colors in the tools or the icons' separation margins and vertical alignment from the text for example)...

So, I'm running the ACBR in dev mode by using ```npm start``` and I did some experiement with the UI.
If we set the ```dir``` attribute on <html> tag, then we have already done most of the task, since flex containers order elements based on this attribute.

Padding and margin issues can be fixed by using CSS logical properties, which set the padding and margin based on the value of the ```dir``` attribute of the element.

You can see in this image that I used '''margin-inline-end''', which is equivalent to ```margin-right``` for the icons on the side menu. When we use CSS, logical properties The direction of margin, padding, etc. is determined based on the ```dir``` attribute.

image

**Can you please look into css logical properites and see if it will be a valid solution?**

So, if we use this solution then we will replace padding/margin with their respective logical properites. Based on what I've seen, we can do this with a few line modifications.


Some elements shouldn't display it's content in RTL, so we can just set ```dir``` for it to be ```auto```.
for example this one:
image

should be like the other LTR languages:
image


I think it should be because this would be the basis for adding RTL support so that ACBR is ready for more potential new RTL languages.

>   * would leaving it like the rest of the languages, left to right, but just using the Arabic translated text (like in the previous images I showed you yesterday) be a good start for now

Displaying Arabic and other RTL languages as if they were written from left to right is definitely not preferable. But I think in the case of ACBR, it's kind of usable; there is nothing that prevents the user from doing the same actions that can be done on LTR interfaces.
So, yes, you can add Arabic to the next release even if RTL support isn't entirely supported.

--
I hope that I've addressed all the points. Let me know if there is anything else to discuss.

**Comment by ahmed-0011** [comment]
@binarynonsense 
Hi, Alvaro
I think I've addressed everything in your previous two comments. Let me know if missed something :)

**Comment by binarynonsense** [comment]
@ahmed-0011 Thanks a lot for taking the time to help me figure things out! And for testing possible changes and letting me know the results of your findings. I like to implement the final code myself in this project, as I use it as a learning experience, but I really appreciate suggestions like these about things to try (including the screenshots, those are really helpful), thanks again for going to the trouble to do that.

Yes, my idea initial is setting the the html tag to <html dir="rtl"> programmatically when an rtl language is loaded (for the screenshots I made, I set it manually to test how that would look), and use the html[dir=rtl] selector in the css files for specific adjustments (I didn't know about 'margin-inline-end' and other variations, I have to read more on that, seems like it cleanly solves the issue with the buttons icon and text margins and their vertically alignment without extra code for each direction) or other 'tricks' like some of the things mentioned [here](https://vmoh.medium.com/conditional-rtl-and-ltr-css-c075074964a3) if needed... In general, I think it's quite doable without adding or changing too many things (especially now that I know how most of the UI needs to look, thanks to you).
 
My main concern is the menu/title bar. I use the package [custom-electron-titlebar](https://www.npmjs.com/package/custom-electron-titlebar) for that and I don't think it has support for RTL out of the box... and, as you can see, when the dir attribute is set to rtl the menus get cut off to the right when expanded if they get close to the edge (same happens normally, the package doesn't really try to fit to both ends, but, as the menu is normally to the left it's not a problem if there aren't too many entries / the window is wide enough.. that's one of the things I least like about it). I have dug into its css/html structure in the past to see how to tweak small things about its look and behaviour to my liking, overriding some of its css from mine, so I may be able to figure out a way to mirror what they do to show the expanded menus and flip them to align to the right instead of the left (and the same for the submenus, they need to appear to the left of the parent instead of the right), but it depends on how the manage that (may be difficult if it's programmatically) and I haven't looked into it yet so I don't really know how hard it will be to 'hack' it without modifying the package's own code (that's something I'd really like to avoid).

As you can see in one of my screenshots, the version text looks good (I assume, the text is switched to "beta1-3.5.0: arabic word") so that will fix itself. As you say, without dir=rtl the weirdness is probably the result of mixing two languages with different directions of writing... (although I'm not sure I understand why the arabic word is put in the middle of the version text when the code is string = arabicword_string + version_string, with the version_string content being the whole "3.5.0-beta1")... I found that odd.

I'm sure I'll have more doubts about specific UI elements and how they should look in rtl mode, so I'll probably bother you again in the future with more questions, hope you don't mind :) But first I'll have to take a look at the title bar and see if changing it is doable without too much trouble. Don't know when I'll be able to do that but I'll let you know if I make some progress.

If you think about anything else that will be needed to make the UI look correct/better for RTL languages or to expand on your suggestions, don't hesitate to tell me. And thanks again for your help.

Álvaro

**Comment by binarynonsense** [comment]
@ahmed-0011 Sorry if you already read this, but I added these as edits to the previous comment yesterday and I just thought you may have read the comment before I did it and may miss them, so just in case I'll move them to their own comment:

Edit: I forgot, seeing that you say that displaying Arabic and other RTL languages as if they were written from left to right is better than nothing in the case of ACBR, I'll add Arabic  to the next releases no matter what, and hopefully add full RTL support at some point if I'm able. 
Edit 2: Please, if you can, try using the program a little like that, with Arabic on but in the default LTR "mode" to see if you catch anything blatantly wrong (maybe due to mixing rtl and ltr without dir=rtl) to make sure I don't release it that way if it's not good enough.

**Comment by binarynonsense** [comment]
Hi @ahmed-0011 

I had a bit of spare time to implement some of the things we talked about. I think it looks better already (see the screenshots, or you can also check it live using npm) but I have a doubt: I've flipped the left and right icons in the toolbar and the mouse click areas so that when you click on the left area of the window, the keyboard's left arrow key or in the toolbar arrow pointing left it will go to the next page (not the previous like in ltr mode) and the opposite for the right (which will match the fact that sliders are mirrored and grow to the left)... is that the expected behavior or the mirroring things and right to left navigation "rules" are not valid when it comes to page-by-page readers?

Didn't have much time to look at the titlebar menu issue with the submenus growing to the right, which is now probably the only major pending issue, but the little I saw was not promising, I still need to think of a way to change the behavior coded in the package without having to make my own version... If I could do it just by overriding their css I'd be happy but I think they calculate the positions programmatically everytime you open a new one and didn't think of a way to bypass that yet, that may be a problem... but I still need to really explore how it works.

image

image

image

**Comment by binarynonsense** [comment]
I think I've managed to find a way to override the title bar code at runtime to change the submenus' positions after they are added, so I can make them expand from right to left. I still need to tweak the css code to rearrange the contents of the menus (text, checkmark, submenu arrow and shortcut text) so they are mirrored, but I'll leave that for another day. I think the hard part is over (don't have time to test it now but my 'hack' seems to work so far) so now I'm more hopeful I'll be able to make the title bar work in rtl mode.

image

**Comment by binarynonsense** [comment]
Ok, had some spare time right now and I've done the tweaks to the css I mentioned yesterday and I think the menu now looks and works as it should in rtl mode. The 'hackish' way I implemented all this may not be too elegant, but I'm happy I found a way to make it work :)

image

@ahmed-0011 if you can give it a try using npm, let me know what you think (or from the screenshots if you can't). And if you can take a look at the question about page changing/area clicking I wrote about a couple comments above (the one with green arrows painted in a screenshot), as that's one is probably the most important remaining decision to make (or my main rtl UI design doubt).

Other than that there are some minor things to fix, like keyboard arrows/gamepad navigation in the tools that allow them and the modals, maybe swap the order of the remove icon and file name in the history... let me know if you see more/think of other things.

**Comment by ahmed-0011** [comment]
Hi!

Sorry for the late response.

Yes, I tried tinkering with the CSS of the main menu, and I too think it is really the most challenging thing in the process of supporting RTL for ACBR.


Yes, I understand your point, but since the arabic word is followed by an english 'word' (-beta1) and <p> in that case is LTR, Then the browser will render it this way (if an element has LTR direction and you write first an arabic word followed by an english word, then the text algorithm will place the english word before the arabic word; as I mentioned before, this doesn't apply to numbers). This is just based on my observations, so I'm not 100% sure :)


By trying ACBR in arabic without changing the `dir` attribute to RTL, the only issue I can see is that arabic text is displayed on the left. So, I think that it's fine if you added arabic to the new release even without full RTL support.


Regarding the reader bar, I think this will be subjective, and it will depend on many factors. For example, YouTube doesn't flip the direction of the control buttons when using the web app in arabic or any other RTL language. The reason for this is probably to make the UI consistent across all languages.

I think we have three options:

1. Mirroring everything like what you did in these screenshots
2. Leave it as it is to make it consistent with different languages.
3. Everything you did in Option 1, but the order of the elements inside the bar stays the same as in LTR languages, and the slider should move from right to left.

I'm not sure if it's a good idea to compare YouTube with ACBR regarding this topic haha. The first time I saw the screenshots, I didn't like the mirroring of the reader bar, but now I like it. So, I guess **option 1** is good enough, imo.

About the screenshot with the "أداة تحويل الملفات" title, I like that the actions for move up/move down/remove from the list are ordered like this, and I think it should be the case for all languages regardless of the writing direction.

**Comment by ahmed-0011** [comment]
I tested the new modifications. Great job, and well done! The final screenshot is exactly the way the main menu and its submenus should look in Arabic.

I think that you've changed some padding values by accident, because the original UI was much better. As you can see in this screenshot, the padding of the elements doesn't look good:

> maybe swap the order of the remove icon and file name in the history... let me know if you see more/think of other things.

This case is similar to the input files listing. I don't think you have to change the order.

**Comment by binarynonsense** [comment]
Thanks for the feedback @ahmed-0011,

- Now that what I call the rtl 'mode' is working as it should the version number strangeness is no longer a problem anyway, as it shows more correctly, I think (beta1-3.5.0 :arabic-word-for-version), but I still find what the browser does curious :) There may be times when that is not what you want... In fact as I typed this I remembered that there's still one problem I noticed yesterday related to this type of things. In the menu, in the open recent files list, the ellipsis dots (...) that are part of the path, to indicate it has been cut short, are moved by the browser when rtl is set and are displayed in the wrong place :( (not sure if I'll be able to fix that)

image

- If you mean the padding of the yellow button and the others in that column, I changed it to make it match how they looked in other languages. For some reason with the font in Arabic the text has a bigger space around the text, which makes the buttons have more apparent total padding, so I had to make an special rule for that... That said, I agree they looked better, so I have just increased the padding a little for all of them (not as big as before the previous change but bigger). I also seem to remember they were bigger at some point so maybe I changed them by accident at some point (also noted that the rounded borders in the buttons that have them only on one side were reversed in rtl mode, so I just fixed that too). Thanks for pointing that out, it helped me find even more issues.

image

- Maybe youtube is not the best reference, as it doesn't work with paginated content and left/right page flow... From what I've read, and what the browser does when setting the document to rtl, sliders in rtl grow from right to left, so the one in the toolbar down below should stay mirrored I assume (same as all other sliders). This means the left and right arrows in the toolbar next to it have to be changed in rtl 'mode', left to go to the next page, right to the previous.... BUT, what if for rtl languages (maybe for all, but I'm a bit adverse to changing things people may be used to if not strictly needed) I change those into up and down arrows, which will mean the same for everyone, up for previous and down for next?. And maybe leave the mouse click the same as with ltr languages (left area prev page, right area next page), maybe add an option in the future to change it in the preference if people request it?
In summary, my current proposal for rtl 'mode' is:
   - keep the toolbar slider mirrored, as it's right now, as it fits the recommended design for rtl UIs
   - change the toolbar left and down arrows for up and down arrows
   - leave the clicking areas as originally, like it was and still is for ltr languages, left area click is prev page, right area click is next
   
Let me know if you like it and if you think an Arabic user would be comfortable using it like that.

Edit: I forgot, I'll leave the remove buttons and other similar things as they are then.

**Comment by ahmed-0011** [comment]
So, for fixing the ellipsis dots's direction, you can simply add the `dir=ltr` attribute to the container of these elements (in the html template). Which makes sense because file paths are mostly written in english (LTR).
I tested this solution, and it works even if the file name is in Arabic, for example.

> * keep the toolbar slider mirrored, as it's right now, as it fits the recommended design for rtl UIs 
> * change the toolbar left and down arrows for up and down arrows
> * leave the clicking areas as originally, like it was and still is for ltr languages, left area click is prev page, right area click is next

I only agree with the first point. for the up/down arrow is presume most users wouldn't like this change, right/left arrow just make scnene becase boosk are read from RTL or LTR.

For point 3, since you are mirroring the reader bar, I think it would be better to leave the clicking left and right regions of the page as you previously described.

**Comment by ahmed-0011** [comment]
Regarding the padding issue, I made this screenshot:

image

On the right, you can see the preferences page with the CSS from the last commit disabled. As you can see, there is no issue with padding when selecting arabic as the language.

On the left, the padding of the button, as you see, is too bad.

So, I'm sure it's not a padding issue; the text in arabic fits perfectly within the element. That's why I don't think padding plays a role here.

Could you please investigate this issue and find the causes?

**Comment by binarynonsense** [comment]
I just tested it on Windows and turns out it is in fact a Linux issue. I've reverted the rtl overrides for the padding, given that most ACBR users are on Windows I prefer that being the version that looks as I expect it (and I didn't like having to use different padding values for rtl, if in the future I need to add other rtl fonts or languages and maybe those don't have that issue and things would look even more different for them). The buttons on Linux (at least on my distro, Kubuntu 22.04) now look chunkier than they should when using Arabic, but they still look decent, so I'll leave it like that:

image

**Comment by ahmed-0011** [comment]
Yes I get email notifications even without the mention.

* Does the library provides a way to add a class or an id to a specific menu item ?

if not here is a quick hack i found:

When the user hover over the file menu, get the "recetly opened" submenu and add the attribute programatically:
```js
// this will get the opened recently element
document.querySelector("[aria-label=ملف]").querySelector(".cet-submenu-item")
```

* Yes, I got your idea, but I think this should be an option that the user can choose in the preferences.
or maybe in the reader bar. Changing this option should determine the behavior of the click left/right region and the left/right arrows. So, yeah, this feature is a bit complex and needs more consideration.

I don't think flipping the page direction always works because the author is presenting his story and envisioning it to be read in a specific direction; at least that's the case in the manga. I'm not sure about other comic books; maybe the same thing applies there. So, yes, they can translate a comic book to arabic and leave the reading direction to be LTR if it isn't possible to make it RTL for some reason.

So, thinking about this, this even supports the need for an option to choose the reading direction, preferably a small element in the reader bar. because it's apparent that language doesn't always determine the reading direction; it's the book.

> If people are already used to comics pages being turned left to right no matter the language that would make the decision easier: leave arrow buttons and click areas / keys the same for all language except maybe the slider

I think you should leave it as it is for now. and maybe later on implement an option for the user to choose his preference. His choice will determine left/right arrow, slider, and page-region-click directions. Also, I think we don't need to mirror the reader bar. It's better to leave the elements in their original position, imo.

(i keep changing my mind about the mirroring, but i think this time im convinced in this one :D)

--

I assumed it would be a linux issue haha :)

**Comment by binarynonsense** [comment]
- No, I can't add ids or anything to anything from the library. and using `document.querySelector("[aria-label=ملف]").querySelector(".cet-submenu-item")` would affect more than those specific items, so it'd break some that do need to be rtl (in that same submenu and others growing from File... as the use the aria-label=File for ). I'd also prefer to make something that works for any rtl language, os using the `aria-label=ملف` for all of them for some reason... 
I'm not on my dev machine right now so I may be missing something, but from memory here's what I'm probably going to try (I'll try to look at it this weekend): For the other changes I made to the titlebar I'm using an observer to detect when a submenu is created, so that will be my starting point.  At first my idea was trying to find a way to detect if I'm in that submenu  (a generic one if possible) and apply the direction=ltr to its children but it has similar problems to what you propose... but I'm thinking I could do something a bit less refined that could fail in other cases but I think would work fine in my case: just check every item when a submenu is created and if the text starts with three dots make it ltr (from the top of my head Idon't think there's any case of a menu item starting with three dots that's not a path). It's a bit ugly but easy to implement and if it works it works... we'll see, I may be missing something doing all this in my head :)

- OK, I think we are getting somewhere :) In the end the solution is to make all the toolbar "ltr" (slider and all) no matter what language is used, and leave the controls as they were before I started adding rtl support. clicking the mouse in the left area, pressing left arrow in the keyboard and clicking the left arrow in the toolbar to go to the previous page, and the opposite to go to the next one. With the idea to maybe add someday (soon if people ask for it or later on whenever I can and there aren't any more urgent things pending) options in the preferences to change the reading direction and flip akll those to the opposite (right now I'm not very keen on adding another button in the toolbar for that, I don't want to have too many more buttons on it and I have the space reserved for a potential feature I'd want to try to add some day, but I'll keep it in my if I implement the option some day)

**Comment by binarynonsense** [comment]
I implemented the solution I detailed yesterday for the menu paths starting with "...", seems to work fine:

image

I've also set the toolbar to ltr, so it always displays the same, and undid the changes to the mouse click areas and arrow keys, so their behavior is also the same again for all languages (left is previous page, right next).

image

If you agree with this being the best solution, I still have one other doubt. Now the optional page number info bubble (View > Show Page Number) and the page number info in the toolbar go in different directions: if, for example, you are in page 3 of 4, the bubble shows "4/3" as it is in rtl mode, and the toolbar shows "3/4"... I think this particular thing should be in rtl mode for both, "4/3", as an rtl reader would expect it that way and it would clash with the rest of the UI otherwise... what do you think? (Also, I'm still torn about mirroring or not the slider, doing it would clash with the left right buttons, but all the other sliders in the app are mirrored...)

image

**Comment by binarynonsense** [comment]
OK, I've been trying to decide what to do and I think we should limit the options to 'all the toolbar as ltr' or 'all the toolbar as rtl'. Then in the preferences I'll add a "Toolbar" section in "User Interface" with a "Direction" select to choose between "Language Direction", "Left to Right" and "Right to Left". And the default being "Language Direction" (it will set the toolbar to rtl for Arabic and ltr for English for example).

I'll also add a "Turn Page Direction" to the Navigation section with the same options ("Language Direction", "Left to Right" and "Right to Left") and this will define what turn to left page and turn to right page will do. The default will be left to right.

 As an Arabic user, would you be OK with all that?

I've added a bunch of new localization entries to support all this (and other things I'm working on for the next release) when I finish it, so if you can, could you update the Arabic file? (as always the new entries are marked as [NEEDS TRANSLATION])

**Comment by ahmed-0011** [comment]
> I'm using an observer to detect when a submenu is created, so that will be my starting point. At first my idea was trying to find a way to detect if I'm in that submenu (a generic one if possible) and apply the direction=ltr to its children but it has similar problems to what you propose...

I guess you can do the following:
1. Get the text of the craeted submenu container and check `submenu_text.includes("Control+H");` or `submenu_text.includes("History")` or, if possible, check the menu item that triggered the creation of this submenu; our target is "Open recently" item.
2. If the the above expression evaluates to true, then add the `dir=ltr` attribute to the submenu, and then get the last item in the submenu, which is "History", and add the `dir=rtl` attribute.

Maybe `MutationObserver` can help in this case?

Unfortunately, your solution in the latest commits for fixing the ... issue works only when the path is truncated with ... at the start; it doesn't apply LTR for other entries that aren't truncated. So, we need a better approach if possible.


Regarding the numbering of the pages, I think it probably should be 3/4 and not 4/3, regardless of the reading direction. For example, in arabic, if we used arabic numeric like this ٤/٣, then yes, it makes sense in this case. But when using 1–9, I think it should always be the current page on the left and the total pages on the right. This is just my opinion. I think we should ignore it for now and leave it as LTR.

I don't think it makes sense to change the slider direction based on the selected language. As we discussed above, the book plays a major role in dictating this. For example, english-translated mangas most of the time will follow the RTL reading direction.

So, basically, the reader should stay LTR for now, and when you have the time to implement the option for reading direction, you probably only need to change the direction/behavior of the slider, the next/previous buttons, and page-region-clicks based on the user's choice. My thoughts are that the reader bar elements should have their original order, and only the behavior should be changed. We can discuss this more when you start implementing the reading direction option, but for now, let's just ignore it because, currently, the reader will stay LTR.



> select to choose between "Language Direction", "Left to Right" and "Right to Left". And the default being "Language Direction" (it will set the toolbar to rtl for Arabic and ltr for English for example, but the user will be able to change it to their liking).

Yes, I like these options. Do you still think that the reader bar should be mirrored when using RTL? for me, I think reversing the slider direction and altering the behavior of the next/previous button and the page regions clicks is enough.

> I'll also add a "Page Turning Direction" to the Navigation section with the same options ("Language Direction", "Left to Right" and "Right to Left") and this will define what 'turn to left page' and 'turn to right page' will do. The default will be left to right.

I feel like I prefer Next page/Previous page. The user will be able to dictate in which direction the page will turn based on the slider, next/previous buttons, and region clicks. So, please reconsider the new translation entries.


I think that the reading direction option should be in the layout submenu in the view menu. that would make it more accessible than if it were in the preferences.

By the way, I was looking to enable the double pages mode, but I can't seem to find the **layout** submenu in the view menu:

image

**Comment by binarynonsense** [comment]
- I didn't think of the possibility of non truncated paths in the history list... In that case I had thought about one of the things you mentioned, checking for the ctrl+h shortcut, as it's the only way I can think of of always identifying that submenu. I'll give it a go (by the way, all this 'hacks' are already based on using a mutation observer to monitor the creation of the submenus).

- Next page/Previous page, which is what I was using before, is too ambiguous in some cases so that made me realize I had to change those. Turn to left/right page fits the visual feedback and iconography better and is always valid whatever you do: left arrow in the toolbar will always mean turn to left, the other way it would change meaning , sometimes will mean next and others prev (I don't like symbols changing meaning). Same thing for defining the hotspots: left will always mean turn to the left page... etc And then you only need to change the page turning direction to control what happens when you press left or right in whatever control or hotspot zone.

- I need to define these things for sure, let me know your opinion but I think this is the way to go unless you find it very wrong, I feel quite good about this right now:
  - I think the toolbar should have just 2 options, one we'll call for 'ltr' and one 'rtl'
  - I think the default (determined by the toolbar direction options I described above) should be "language direction", so by default the Arabic would be rtl and English ltr. This follow all the guidelines I've read about rtl language UI design. The layout needs to be mirrored in rtl because the toolbar is designed to be read from left to right, if the UI is changed to rtl and all other elements shift focus and importance from left to now right it should too. 
  - There is an exception to the above: the toolbar slider. I think it should be determined by the "page turning direction" (which is what indicates the direction the document grows towards). This (plus the renaming to 'turning to left/right page") makes everything cohesive and logical.
  - My only doubt is the page bubble and the toolbar page number text. Could you try to find programs or documents using that format (number/number) in Arabic to see what they do? In the mean time I may force it to ltr always in both, 3/4 for example.

- I feel like the page turning direction is something the user would change once or twice, and it's not related to how you view the pages but how the UI elements translate to actions, so it fits better in the UI preferences. You are not going to be changing it based on the book, no matter if it's manga or an american one (being digital the original page binding direction doesn't matter), you'll use the setting you are more comfortable with and use it always like that. It would be a different matter if there was an horizontal continuous mode, but that is not going to happen.  But I may change my opinion in the future, it's not the first time :)

- There is no layout submenu or double pages mode :) and may never be. There are some entries in the localization files related to that, and other page layouts, to be ready just in case I try and manage to add them in the future (like double page and continuous mode, with continuous mode being the less likely to get done for a bunch of different reasons, but we'll see).

**Comment by binarynonsense** [comment]
About adding the direction config to the view > layout menu, I think I'm going to do that. This will help me future-proof things in case I add a double page mode, as that will also serve to determine which page goes to the left and which to the right. So I'm gonna move and rename the entry in the localization files and add a view>layout>page turning direction menu, with left to right (default) and right to left options... and maybe remember what it's set to per book in the history (like the zoom)

**Comment by binarynonsense** [comment]
I just pushed a commit that fixes the non truncated paths in history menu list not being set to ltr. Implemented what I described and seems to work fine from the few tests I've made.

**Comment by binarynonsense** [comment]
OK, I've implemented and pushed everything I mentioned and I think I'm happy with things so, unless you have some major objection, I'm gonna make the decision to choose this configuration and leave it like this (as long as there isn't anything really wrong, we have to decide to stop at some point, at least for now :))

There are now 2 new configuration options: 
  -  one in the preferences (appearance>toolbar>direction) to set the general direction of the toolbar's layout (default is infer from language)
  - another in the menu bar (view>layout>pages direction) that determines what happens when you input turn to left page and turn to right page (using any of the input methods) and the direction of the toolbar's slider.
  
With all this, even if a user is not comfortable with the default look and behavior, I think there are enough configuration options to change them to their liking.

**Comment by ahmed-0011** [comment]
Hi, Alvaro

I like the new changes and improvements (especially the small page direction button on the reader bar), and I agree that it's a good idea to move the show UI items, such as "Show Clock" etc., because it makes sense to put them under the Layout submenu besides page direction. In addition, adding the page direction there makes it more accessible because I feel that some users will probably change it based on the type of book. I also like the idea of saving the direction of the books in history, and I find it interesting.


Unfortunately, it seems that the paths in the "Recently Opened" submenu are still aligned to the right. Please confirm if it's the same case for you.

Regarding changing the Next Page/Pervios page into turn to left/turn to right, now I understand your point, and I think it's a good point to consider.

So far, I like the implementation of the page direction feature, but here is a thing rearding the slider: In LTR mode, the reader slider progress is represented by growing the yellow portion on the left of the slider tick. So in RTL mode, I think the reader slider should have the yellow portion start from the right of the tick and grow with moving to the left, which will make it consistent with how the progress of reading pages should be represented. Currently, the yellow full progress means you are on the first page, and the gray full progress means you are on the last page. Basically, you just need to switch the colors for the slider and the container's background.

For the page bubble and the toolbar page number text, I'm leaning toward leaving them as LTR, but I'll look more into it, and if I find something, I will let you know.

---

The page direction submenu appears on top of the title bar when using arabic, Please check what is causing this issue:

image


I suggest that you set the `lang` attribute for the `html` element because languages with the same direction might have different expectations. This property will help to customize the behavior of some parts of ACBR based on the language. (I don't think it has any uses now, but maybe later on.)

**Comment by binarynonsense** [comment]
Hi @ahmed-0011 

- yes, the recently opened items are aligned to the right in the Arabic UI, as they are to the left in the ltr languages. I'm trying to maintain the mirroring principle as much as possible so I think it's that's the way to go as the rest of the menu follows that principle.
- you are right, I didn't noticed that the slider yellow part is always to the left no matter what, I'll look into it
- good catch, I probably didn't tested the page direction in Arabic so I didn't noticed that... I'm guessing it has to do something with being a third level submenu, just guessing, as I don't thing there are any other in the UI.
- Adding the lang attribute is in my TODO list, but haven't done it so far because I didn't think it mattered so far. I'll probably do it anyway soon.

Thanks for the feedback!

**Comment by ahmed-0011** [comment]
> yes, the recently opened items are aligned to the right in the Arabic UI, as they are to the left in the ltr languages. I'm trying to maintain the mirroring principle as much as possible so I think that's the way to go, as the rest of the menu follows that principle.

Since file paths most of the time are in "English", I think they probably should be displayed in LTR.

For example:
`C:\Users\user1\Desktop\...` OR `\home\user1\...`

in previous comments, we discussed that the "Recently Opened" items should all be in LTR and not only the ones truncated with ...

That's why I expected them to be in LTR... Overall, they are not bad when displaying them in RTL; I just think that maybe displaying them in LTR would be a better choice. In this way, you will be consistent with displaying truncated paths in LTR.

**Comment by binarynonsense** [comment]
OK, haven't tested it much but I think the third level submenus wrong vertical position and the toolbar yellow part not mirroring when rtl page direction is set are now fixed.

**Comment by ahmed-0011** [comment]
So, I tested ACBR with the recent changes, and they work as expected.

While testing the app, I found some minor issues with the padding of the hint icons for the input titles (`margin-left` is used instead of `margin-start`):

image

**Comment by binarynonsense** [comment]
Thanks! I'll fix that.

**Comment by binarynonsense** [comment]
> So, I tested ACBR with the recent changes, and they work as expected.
> 
> While testing the app, I found some minor issues with the padding of the hint icons for the input titles (`margin-left` is used instead of `margin-start`)

Fixed.

I also fixed a regression, at least on linux the tool menu icons where not aligning properly vertically. They now do again:

image
